File: scraper.py
import os
import re
import time
import csv
import requests
from bs4 import BeautifulSoup
from concurrent.futures import ThreadPoolExecutor

# Constants
BASE_URL = "https://papers.nips.cc"
SAVE_DIRECTORY = "E:/Python_Scraper/"
CSV_FILE_PATH = os.path.join(SAVE_DIRECTORY, "output.csv")
START_YEAR = 2019
END_YEAR = 2023
THREAD_SLEEP_TIME = 2  # Sleep time in seconds
MAX_THREADS = 10
TIMEOUT = 20  # Timeout for HTTP requests


def create_directory(path):
    """Create a directory if it doesn't exist."""
    os.makedirs(path, exist_ok=True)


def create_csv():
    """Create CSV file and write the header if it doesn't exist."""
    create_directory(SAVE_DIRECTORY)
    if not os.path.exists(CSV_FILE_PATH):
        with open(CSV_FILE_PATH, mode="w", newline="", encoding="utf-8") as file:
            writer = csv.writer(file)
            writer.writerow(["Year", "Title", "Authors", "Abstract", "PDF Link"])


def fetch_page(url):
    """Fetch page content with retry mechanism."""
    for _ in range(3):  # Retry up to 3 times
        try:
            response = requests.get(url, timeout=TIMEOUT)
            if response.status_code == 200:
                return response.text
        except requests.RequestException as e:
            print(f"Error fetching {url}: {e}")
        time.sleep(THREAD_SLEEP_TIME)  # Rate limiting
    return None


def extract_year_from_text(text):
    """Extract year from a given text using regex."""
    match = re.search(r"(\d{4})", text)
    return int(match.group(1)) if match else None


def process_year_page(year_url, year):
    """Extract paper links from a given year's page and process them."""
    html = fetch_page(year_url)
    if not html:
        return

    soup = BeautifulSoup(html, "html.parser")
    paper_links = soup.select("a[href^='/paper/']")  # Fixing paper link selector

    with ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
        for paper in paper_links:
            paper_url = BASE_URL + paper["href"]
            executor.submit(process_paper_page, paper_url, year)


def process_paper_page(paper_url, year):
    """Extract metadata and PDF link from a paper's page."""
    html = fetch_page(paper_url)
    if not html:
        return

    soup = BeautifulSoup(html, "html.parser")

    title = soup.find("h4").text.strip() if soup.find("h4") else "N/A"
    authors = ", ".join([i.text.strip() for i in soup.find_all("i")])
    abstract_text = soup.find("p").text.strip() if soup.find("p") else "N/A"

    pdf_link_element = soup.select_one("a[href$='.pdf']")
    pdf_url = BASE_URL + pdf_link_element["href"] if pdf_link_element else "N/A"

    save_metadata_to_csv(year, title, authors, abstract_text, pdf_url)

    if pdf_url != "N/A":
        download_pdf(pdf_url, year)


def download_pdf(pdf_url, year):
    """Download and save the PDF."""
    pdf_name = pdf_url.split("/")[-1]
    year_folder = os.path.join(SAVE_DIRECTORY, str(year))
    create_directory(year_folder)
    pdf_path = os.path.join(year_folder, pdf_name)

    try:
        response = requests.get(pdf_url, timeout=TIMEOUT, stream=True)
        with open(pdf_path, "wb") as file:
            for chunk in response.iter_content(chunk_size=8192):
                file.write(chunk)
        print(f"Downloaded: {pdf_name} to {year_folder}")
    except requests.RequestException as e:
        print(f"Error downloading {pdf_name}: {e}")


def save_metadata_to_csv(year, title, authors, abstract_text, pdf_url):
    """Save extracted data to a CSV file."""
    with open(CSV_FILE_PATH, mode="a", newline="", encoding="utf-8") as file:
        writer = csv.writer(file)
        writer.writerow([year, title, authors, abstract_text, pdf_url])


def main():
    """Main function to scrape research papers from NeurIPS."""
    create_csv()

    homepage_html = fetch_page(BASE_URL)
    if not homepage_html:
        print("Failed to load homepage.")
        return

    soup = BeautifulSoup(homepage_html, "html.parser")
    year_links = soup.select("a[href^='/paper_files/paper/']")

    with ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
        for link in year_links:
            extracted_year = extract_year_from_text(link.text)
            if extracted_year and START_YEAR <= extracted_year <= END_YEAR:
                year_url = BASE_URL + link["href"]
                executor.submit(process_year_page, year_url, extracted_year)


if __name__ == "__main__":
    main()

File: Scraper.java
package java_proj;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.concurrent.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Scraper {

    // Base website URL
    private static final String WEBSITE_URL = "https://papers.nips.cc";
    // Directory where data will be saved
    private static final String SAVE_DIRECTORY = "E:/Java_Scraper/";
    // CSV file to store paper details
    private static final String CSV_FILE_PATH = SAVE_DIRECTORY + "output.csv";
    
    // Scraper configuration
    private static final int REQUEST_TIMEOUT = 200000; // Max time for a request (ms)
    private static final int MAX_THREADS = 20; // Number of threads for parallel execution
    private static final int STARTING_YEAR = 2019;
    private static final int ENDING_YEAR = 2023;
    private static final int THREAD_SLEEP_TIME = 2000; // Sleep time between requests (ms)

    public static void main(String[] args) {
        ExecutorService threadPool = Executors.newFixedThreadPool(MAX_THREADS);
        createCSVFile(); // Create CSV file if not exists

        try {
            // Fetch the homepage HTML
            Document homepage = Jsoup.connect(WEBSITE_URL).timeout(REQUEST_TIMEOUT).get();
            Elements yearLinks = homepage.select("a[href^=/paper_files/paper/]");

            for (Element link : yearLinks) {
                String linkText = link.text();
                Integer extractedYear = extractYearFromText(linkText);
                
                // Process only the years in the given range
                if (extractedYear != null && extractedYear >= STARTING_YEAR && extractedYear <= ENDING_YEAR) {
                    String yearPageUrl = WEBSITE_URL + link.attr("href");
                    threadPool.submit(() -> processYearPage(yearPageUrl, extractedYear));
                }
            }
        } catch (IOException e) {
            System.err.println("Error connecting to homepage: " + e.getMessage());
        }
        threadPool.shutdown();
    }

    // Extracts year from a text string
    private static Integer extractYearFromText(String inputText) {
        Pattern pattern = Pattern.compile("(\\d{4})");
        Matcher matcher = pattern.matcher(inputText);
        if (matcher.find()) {
            return Integer.parseInt(matcher.group(1));
        }
        return null;
    }

    // Processes a specific year's page and fetches paper links
    private static void processYearPage(String yearPageUrl, int year) {
        try {
            Thread.sleep(THREAD_SLEEP_TIME);
            Document yearPage = Jsoup.connect(yearPageUrl).timeout(REQUEST_TIMEOUT).get();
            Elements paperLinks = yearPage.select("ul.paper-list li a[href$=Abstract-Conference.html]");

            for (Element paper : paperLinks) {
                String paperPageUrl = WEBSITE_URL + paper.attr("href");
                processPaperPage(paperPageUrl, year);
            }
        } catch (IOException | InterruptedException e) {
            System.err.println("Error processing year page: " + e.getMessage());
        }
    }

    // Processes a specific paper's page to extract metadata and download PDF
    private static void processPaperPage(String paperPageUrl, int year) {
        try {
            Thread.sleep(THREAD_SLEEP_TIME);
            Document paperPage = Jsoup.connect(paperPageUrl).timeout(REQUEST_TIMEOUT).get();

            String title = paperPage.selectFirst("h4").text();
            String authors = paperPage.select("i").text();
            String abstractText = paperPage.selectFirst("p").text();
            
            // Extract PDF link
            Element pdfDownloadLink = paperPage.selectFirst("a[href$=Paper-Conference.pdf]");
            String pdfUrl = (pdfDownloadLink != null) ? WEBSITE_URL + pdfDownloadLink.attr("href") : "N/A";

            // Save details to CSV
            writeMetadataToCSV(year, title, authors, abstractText, pdfUrl);
            
            // Download PDF if available
            if (pdfDownloadLink != null) {
                downloadPdfFile(pdfUrl, year);
            }
        } catch (IOException | InterruptedException e) {
            System.err.println("Error processing paper page: " + e.getMessage());
        }
    }

    // Downloads the PDF file from the given URL
    private static void downloadPdfFile(String pdfUrl, int year) {
        String pdfFileName = pdfUrl.substring(pdfUrl.lastIndexOf('/') + 1);
        String yearFolderPath = SAVE_DIRECTORY + year + "/";

        try {
            Files.createDirectories(Paths.get(yearFolderPath));
            URL fileUrl = new URL(pdfUrl);
            HttpURLConnection connection = (HttpURLConnection) fileUrl.openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(REQUEST_TIMEOUT);
            connection.setReadTimeout(REQUEST_TIMEOUT);

            Thread.sleep(THREAD_SLEEP_TIME);
            try (InputStream inputStream = connection.getInputStream();
                 FileOutputStream outputStream = new FileOutputStream(yearFolderPath + pdfFileName)) {
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = inputStream.read(buffer)) != -1) {
                    outputStream.write(buffer, 0, bytesRead);
                }
            }
            System.out.println("Downloaded: " + pdfFileName + " to " + yearFolderPath);
        } catch (IOException | InterruptedException e) {
            System.err.println("Error downloading PDF: " + e.getMessage());
        }
    }

    // Creates the CSV file to store paper details
    private static void createCSVFile() {
        File csvFile = new File(CSV_FILE_PATH);
        try {
            Files.createDirectories(Paths.get(SAVE_DIRECTORY));
            if (!csvFile.exists()) {
                try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(csvFile), "UTF-8"))) {
                    writer.write("Year,Title,Authors,Abstract,PDF Link\n");
                    writer.flush();
                }
            }
        } catch (IOException e) {
            System.err.println("Error creating CSV file: " + e.getMessage());
        }
    }

    // Writes extracted data to CSV
    private static synchronized void writeMetadataToCSV(int year, String title, String authors, String abstractText, String pdfUrl) {
        try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(CSV_FILE_PATH, true), "UTF-8"))) {
            title = title.replace("\"", "'");
            authors = authors.replace("\"", "'");
            abstractText = abstractText.replace("\"", "'");
            writer.write(String.format("\"%d\",\"%s\",\"%s\",\"%s\",\"%s\"\n", year, title, authors, abstractText, pdfUrl));
            writer.flush();
        } catch (IOException e) {
            System.err.println("Error writing to CSV file: " + e.getMessage());
        }
    }
}
